### PR TITLE
Support specifying distance between categorical choices in `TPESampler`

### PR DESCRIPTION
**Pull request by knshnb**
## Motivation

Previously, searching a large categorical space was diffucult because all choices in `CategoricalDistribution` were handled equally.
This PR supports specifying any distance function between categorical choices in `TPESampler`, which enables the sampler to take into account the user-defined structure of the search space for efficient optimization. By this, we can handle many different structures, such as permutations, embeddings, graph nodes, etc.

Toy example: find a ground-truth permutation from all permutations using hamming distance.
```python
import itertools
import random

import numpy as np

import optuna


def hamming_distance(perm1, perm2):
    return float((np.array(perm1) != np.array(perm2)).sum())


n = 6
perms = list(map(list, itertools.permutations(range(n))))
groundtruth_perm = random.choice(perms)


def hamming_objective(trial):
    perm = trial.suggest_categorical("perm", perms)
    return hamming_distance(groundtruth_perm, perm)


study = optuna.create_study(
    sampler=optuna.samplers.TPESampler(categorical_distance_func={"perm": hamming_distance})
)
study.optimize(hamming_objective, n_trials=100)
optuna.visualization.plot_optimization_history(study).show()
```

We can quickly find the optimal solution from 6! (=720) choices:
image

RandomSampler:
image

## Description of the changes

Added a `categorical_distance_func` argument to `TPESampler`.

### CODE REVIEW

File: optuna/samplers/_tpe/sampler.py [hunk 1]
```diff
@@ -228,7 +229,20 @@ def objective(trial):
                 Added in v3.0.0 as an experimental feature. The interface may change in newer
                 versions without prior notice.
                 See https://github.com/optuna/optuna/releases/tag/v3.0.0.
+        categorical_distance_func:
+            A dictionary of distance functions for categorical variables. The key is the name of
+            the categorical variable and the value is a distance function that takes two
+            :class:`~optuna.distributions.CategoricalChoiceType` s and returns a :obj:`float`
+            value. The distance function must return a non-negative value.
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:236]:
```suggestion
        categorical_distance_func:
            A dictionary of distance functions for categorical parameters. The key is the name of
            the categorical parameter and the value is a distance function that takes two
            :class:`~optuna.distributions.CategoricalChoiceType` s and returns a :obj:`float`
            value. The distance function must return a non-negative value.
```

File: optuna/samplers/_tpe/sampler.py [hunk 2]
```diff
@@ -228,7 +229,20 @@ def objective(trial):
...
 
+            While categorical choices are handled equally by default, this option allows users to
+            specify prior knowledge on the structure of categorical variables. When specified,
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:239]:
```suggestion
            specify prior knowledge on the structure of categorical parameters. When specified,
```

File: tests/samplers_tests/tpe_tests/test_parzen_estimator.py [hunk 3]
```diff
@@ -215,13 +217,48 @@ def test_calculate_shape_check(
...
     )
     assert len(mpe._mixture_distribution.weights) == max(len(mus) + int(prior), 1)
 
 
+@pytest.mark.parametrize("mus", (np.asarray([]), np.asarray([0.4]), np.asarray([-0.4, 0.4])))
+@pytest.mark.parametrize("prior_weight", [1.0, 0.01, 100.0])
+@pytest.mark.parametrize("prior", (True, False))
+@pytest.mark.parametrize("magic_clip", (True, False))
+@pytest.mark.parametrize("endpoints", (True, False))
+@pytest.mark.parametrize("multivariate", (True, False))
```
Comment (HideakiImamura) [inline tests/samplers_tests/tpe_tests/test_parzen_estimator.py:233]: It seems that these parameters are not related to the categorical distance. Can we fix them?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the great PR. Basically, LGTM. I have several minor comments.

File: tests/samplers_tests/tpe_tests/test_parzen_estimator.py (hunk 3 above)
Comment (knshnb) [reply tests/samplers_tests/tpe_tests/test_parzen_estimator.py]: That makes sense. I removed three parameters, "magic_clip", "endpoints", and "multivariate", which are not used in `_calculate_categorical_distributions`.

**Review by HideakiImamura — APPROVED** [review]
Thanks for your swift actions. LGTM!

**Review by contramundum53 — APPROVED** [review]
LGTM!
